### PR TITLE
Implement table loading state

### PR DESCRIPTION
**Pull request by michael-yx-wu**
#### Checklist
- [ ] Include tests
- [ ] Update documentation

#### Changes proposed in this pull request:
- Regional loading options (e.g. all column headers, all row headers, or all body cells)
- Individual cells and header cells can show loading state for more granular control
- Fade in content when transitioning from loading state

### CODE REVIEW

**Comment by giladgray** [comment]
commit history is very bizarre here because it includes lots of obsolete changes since #338. @michael-yx-wu would be great if you could clean that up so the history is more legible, especially given that you're changing 20+ files. I'd like to be able to sanely read this commit-by-commit.

**Comment by michael-yx-wu** [comment]
Sure I'll rebase and squash down commits that are no longer relevant!

**Comment by giladgray** [comment]
you might have an easier time starting fresh with copy-paste, but i'll leave it up to you.

File: packages/table/examples/tableLoadingExample.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import * as React from "react";
+
+import { Switch } from "@blueprintjs/core";
+import BaseExample from "@blueprintjs/core/examples/common/baseExample";
+
+import { Cell, Column, RegionCardinality, RowHeaderCell, Table, TableLoadingOption } from "../src";
+
+interface IFuelConsumption {
+    country: string;
+    [ index: number ]: any;
```
Comment (giladgray) [inline packages/table/examples/tableLoadingExample.tsx]: pretty sure json keys are always strings, by definition.

File: packages/table/src/cell/editableCell.tsx [hunk 2]
```diff
@@ -44,27 +48,39 @@ export class EditableCell extends React.Component<IEditableCellProps, {}> {
...
 
     public render() {
-        const { className, value, intent, onCancel, onChange, onConfirm, style, tooltip } = this.props;
+        const { className, value, intent, isLoading, onCancel, onChange, onConfirm, style, tooltip } = this.props;
+
+        const classes = classNames(
+            CELL_CLASSNAME,
+            Classes.intentClass(intent),
+            {
+                "pt-loading": isLoading,
```
Comment (giladgray) [inline packages/table/src/cell/editableCell.tsx]: `[Classes.LOADING]`?

File: packages/table/src/column.tsx [hunk 3]
```diff
@@ -21,6 +24,13 @@ export interface IColumnProps extends IColumnNameProps, IProps {
     id?: string | number;
 
     /**
+     * A single value or set of `ColumnLoading` options. If non-null, loading
```
Comment (giladgray) [inline packages/table/src/column.tsx]: what's this `ColumnLoading` name? don't see it defined anywhere

Comment (giladgray) [inline packages/table/src/column.tsx]: also don't say single value cuz it's always gotta be an array

File: packages/table/src/common/loadableContent.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as React from "react";
+
+import { ILoadable } from "./loading";
+
+export interface ILoadableContentProps extends ILoadable {
+
```
Comment (giladgray) [inline packages/table/src/common/loadableContent.tsx]: remove whitespace

File: packages/table/src/common/loadableContent.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    private skeletonClassName = "pt-skeleton";
+    private skeletonLength: number;
+
+    public constructor(props: ILoadableContentProps) {
+        super(props);
+        this.skeletonLength = props.variableLength ? 100 - Math.floor(Math.random() * 4) * 5 : 100;
+    }
+
+    public render() {
+        const { isLoading } = this.props;
```
Comment (giladgray) [inline packages/table/src/common/loadableContent.tsx]: don't bother deconstructing if you're going to use it exactly once.

File: packages/table/src/common/loadableContent.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        if (isLoading) {
+            return <div className={`${this.skeletonClassName} ${this.skeletonClassName}-${this.skeletonLength}`} />;
```
Comment (giladgray) [inline packages/table/src/common/loadableContent.tsx]: this seems like the perfect opportunity for inline styles... would also allow many more "random" lengths.

File: packages/table/src/common/loadableContent.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Show a variable length skeleton when rendering the loading state.
+     * @default false
+     */
+    variableLength?: boolean;
+}
+
+export class LoadableContent extends React.Component<ILoadableContentProps, {}> {
+    private skeletonClassName = "pt-skeleton";
```
Comment (giladgray) [inline packages/table/src/common/loadableContent.tsx]: what is this field for? how about just using `Classes.SKELETON`?

File: packages/table/src/common/loading.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as React from "react";
+
+export interface ILoadable {
```
Comment (giladgray) [inline packages/table/src/common/loading.tsx]: export does not match filename. not convinced this needs its own file, either. why not merge with LoadableContent?

File: packages/table/src/common/loading.tsx [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    /**
+     * Show an animated loading animation.
+     * @default false
+     */
+    isLoading?: boolean;
```
Comment (giladgray) [inline packages/table/src/common/loading.tsx]: i think i'd prefer to call it simply `loading` [like we're doing with buttons](https://github.com/palantir/blueprint/pull/349/files#diff-76186512bf8bfa4ccfa05cb3d9372687R30)

File: packages/table/typings.json [hunk 10]
```diff
@@ -1,6 +1,8 @@
 {
   "name": "@elements/blueprint-table",
-  "dependencies": {},
+  "dependencies": {
+    "lodash": "registry:npm/lodash#4.0.0+20161015015725"
```
Comment (giladgray) [inline packages/table/typings.json]: __whoa__ is this actually used? it doesn't appear in package.json.

File: packages/table/src/tableBody.tsx [hunk 11]
```diff
@@ -5,24 +5,30 @@
...
+import { ColumnLoadingOption, Regions } from "./regions";
 
-export interface ITableBodyProps extends ISelectableProps, IRowIndices, IColumnIndices, IProps {
+export interface ITableBodyProps extends IColumnIndices, ILoadable, IProps, IRowIndices, ISelectableProps {
     /**
      * A cell renderer for the cells in the body.
      */
     cellRenderer: ICellRenderer;
 
+    getColumnProps: (columnIndex: number) => IColumnProps;
```
Comment (giladgray) [inline packages/table/src/tableBody.tsx]: **:fire: adding a new required prop is a significant API break.** what is the purpose of this prop? at the very least it's going to need some clear documentation.

File: packages/table/src/cell/editableCell.tsx (hunk 2 above)
Comment (michael-yx-wu) [inline packages/table/src/cell/editableCell.tsx]: omg hooray for this ES6 function. will

**Comment by giladgray** [comment]
superseded by #435 etc

**Pull request closed without merging**